### PR TITLE
[Testcase Refactoring] Label test_c10d_torchcomms classes with HardwareClassification

### PR DESCRIPTION
**Pull request by twthighd**
## Summary
This PR adds `hw_classification` labels to the `TestCase` subclasses in `test/distributed/test_c10d_torchcomms.py`, drops the hard-coded `only_for` whitelist on the ACCELERATOR classes, and — per review feedback — generalizes the device-dependent tests to resolve backends through the framework registry, splitting NCCL-specific tests into dedicated CUDA classes.

## Changes
- **`test/distributed/test_c10d_torchcomms.py`**:
  - `TestC10dTorchCommsBasic` -> `ACCELERATOR` (instantiated via `instantiate_device_type_tests(..., allow_xpu=True)`; verifies collective correctness across device types). Generalized the `new_group` backend-qualification tests to use `self.backend(torch.device(device).type)` instead of hard-coded `"nccl"` / `"cuda:nccl"` . Removed the now-unused `_requires_cuda` helper.
  - `TestC10dTorchCommsInitAutoQualify` -> `ACCELERATOR` (was `CUDA`, `only_for=["cuda"]`). Generalized `_init_pg` to resolve `device_type` via `cls.device_type` (handling `callable` + `"privateuse1"` alias) and `backend` via `cls.backend(device_type)`, using `torch.get_device_module(device_type).set_device(rank)` instead of `torch.cuda.set_device`. `test_default_pg_has_cuda_backend` keeps its name but uses `_get_backend(torch.device(torch.device(device).type))`. `test_bound_device_id_is_set` asserts `.type == torch.device(device).type`.
  - New `TestC10dTorchCommsBasicCUDA` (`CUDA`, `only_for=["cuda"]`): NCCL-specific `pg_options` tests (`test_new_group_with_pg_options`, `test_new_group_sequential_pg_options_produce_distinct_groups`) using `dist.ProcessGroupNCCL.Options`.
  - New `TestC10dTorchCommsInitAutoQualifyCUDA` (`CUDA`, `only_for=["cuda"]`): `test_new_group_nccl_lazy_builds_per_peer_group` — `nccl-lazy` is registered for CUDA devices only (`devices=["cuda"]`); keeps the original CUDA `_init_pg` (`backend="nccl"`, `cuda:{rank}`).
  - `TestC10dTorchCommsMixedBackends` / `TestTorchCommsHandlesBackend` / `TestC10dTorchCommsNewGroupHelper` / `TestC10dGroupNameHashSalt` / `TestC10dTorchCommsDestroyDedup` / `TestC10dTorchCommsBackendConfig` -> `CUDA` (`only_for=["cuda"]`). These test TorchComms-specific c10d logic and carry hard-coded `"cuda"` / `backend="nccl"` values.
  - Drop `only_for=["cpu","cuda","xpu"]` from the ACCELERATOR classes' `instantiate_device_type_tests` calls (keep `allow_xpu=True`).

## Motivation
The original file had no `hw_classification` labels, so `--hw-classification` filtering silently dropped every class. Labeling them makes the classes collectible under hardware-classification-filtered CI runs.

The ACCELERATOR classes (`TestC10dTorchCommsBasic`, `TestC10dTorchCommsInitAutoQualify`) are generalized to resolve backends through `C10dTorchCommsTestBase.backend()` (which itself is generalized in #194152 to use `c10d.get_default_backend_for_device`), so out-of-tree accelerators that register their comm backend via `Backend.register_backend` are recognized. NCCL-specific tests (`ProcessGroupNCCL.Options`, `nccl-lazy`) are split into dedicated `CUDA` classes to keep the ACCELERATOR classes clean.

The 6 CUDA classes are labeled `CUDA` (not `GENERIC`) because they test TorchComms-specific c10d logic and carry hard-coded `"cuda"` / `backend="nccl"` values. 

## Test Plan
- Verified on an 8-card accelerator backend (no CUDA,):

```bash
# ACCELERATOR class — accelerator variant resolves to hccl, skips cleanly
python test/distributed/test_c10d_torchcomms.py TestC10dTorchCommsBasicPRIVATEUSE1
# Ran 35 tests — OK (skipped=35, "torchcomms hccl backend is not available", 0 errors)
```
- `lintrunner` reports no lint issues. 62/62 tests preserved (no test loss — only class reorganization).

## Notes
- **Depends on #194152** (framework-level `C10dTorchCommsTestBase.backend()` / `_skip_if_backend_unavailable` / `_init_pg` generalization). The ACCELERATOR classes' generalized `_init_pg` and `backend()` calls rely on #194152's `get_default_backend_for_device` + `TORCHCOMM_HAS_HCCL` to resolve and skip correctly on non-CUDA backends. Without #194152, `backend(<accelerator device type>)` returns `"gloo"` (hard-coded fallback) and tests would false-pass on the gloo/CPU path.
- `skip_if_lt_x_gpu` device decoupling (letting that helper recognize out-of-tree accelerators) is tracked by pytorch/pytorch#187650 — an independent framework-level change to `common_distributed.py`, orthogonal to this PR. This PR keeps `@skip_if_lt_x_gpu` unchanged.
- This is part of the test case refactoring initiative tracked in #185590.

### CODE REVIEW

**Comment by twthighd** [comment]
@wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

**Comment by jishuangfeng** [comment]
Nice cleanup — removing `only_for=["cpu","cuda","xpu"]` to let `instantiate_device_type_tests` auto-discover PrivateUse1 backends aligns with the device-agnostic direction in #185590.

One note: `allow_xpu=True` is now somewhat redundant without `only_for`(its purpose was to allow XPU beyond the `only_for` list). Worth confirming it doesn't cause double-instantiation or unexpected behavior, otherwise consider droppi ng it as well.

LGTM modulo that check.

**Comment by twthighd** [comment]
> Nice cleanup — removing `only_for=["cpu","cuda","xpu"]` to let `instantiate_device_type_tests` auto-discover PrivateUse1 backends aligns with the device-agnostic direction in #185590.
> 
> One note: `allow_xpu=True` is now somewhat redundant without `only_for`(its purpose was to allow XPU beyond the `only_for` list). Worth confirming it doesn't cause double-instantiation or unexpected behavior, otherwise consider droppi ng it as well.
> 
> LGTM modulo that check.

Kept allow_xpu=True since XPU needs explicit opt-in and isn't auto-configured. It's just an opt-in switch, so it won't cause any duplicate instantiation issues.

**Comment by jishuangfeng** [comment]
> > Nice cleanup — removing `only_for=["cpu","cuda","xpu"]` to let `instantiate_device_type_tests` auto-discover PrivateUse1 backends aligns with the device-agnostic direction in #185590.
> > One note: `allow_xpu=True` is now somewhat redundant without `only_for`(its purpose was to allow XPU beyond the `only_for` list). Worth confirming it doesn't cause double-instantiation or unexpected behavior, otherwise consider droppi ng it as well.
> > LGTM modulo that check.
> 
> Kept allow_xpu=True since XPU needs explicit opt-in and isn't auto-configured. It's just an opt-in switch, so it won't cause any duplicate instantiation issues.

Thanks for the clarification — that makes sense.
LGTM.

File: test/distributed/test_c10d_torchcomms.py [hunk 1]
```diff
@@ -573,6 +577,7 @@ def test_empty_parts_are_skipped(self):
 
 
 class TestC10dTorchCommsNewGroupHelper(TestCase):
+    hw_classification = HardwareClassification.GENERIC
```
Comment (wjlFlyer) [inline test/distributed/test_c10d_torchcomms.py]: The classification label for these test cases are "generic", meaning they are device-agnostic. So, if we are using a third-party backend (e.g. NPU), does this means that we need to filter out these cases (for example, by configuring a blacklist via yaml or json file?

Comment (wjlFlyer) [reply test/distributed/test_c10d_torchcomms.py]: Furthermore，I find that there're hardcode device information in these test cases, such as "cuda". So, if we set classification label as "generic", these cases may run failed on other device (e.g. xpu, mps). Maybe the CI workflow will failed.

Comment (wjlFlyer) [reply test/distributed/test_c10d_torchcomms.py]: Since TorchComms is also experimental library, and it's mainly used on gpu and cuda. So I think we should filter other device, such as xpu,mps, privateUse1.

Comment (wjlFlyer) [reply test/distributed/test_c10d_torchcomms.py]: In general, maybe it's better to change "generic" to "cuda".

Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py]: Thanks for the review — you're right, and I'll change all `GENERIC` classes in this file to `CUDA`.

Will push the update shortly.

**Review by wjlFlyer — APPROVED** [review]

File: test/distributed/test_c10d_torchcomms.py [hunk 2]
```diff
@@ -22,6 +23,8 @@
 
 @unittest.skipIf(not _TORCHCOMM_AVAILABLE, "TorchComms is not installed")
 class TestC10dTorchCommsBasic(C10dTorchCommsTestBase):
```
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:25]: Hi @twthighd, I think there are so many works we need to do in `C10dTorchCommsTestBase` and its inherited class, right? Or this test class can not be tested correctly in PrivateUse1-based Accelerator.

Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:25]: We need to refactor `backend`, `_skip_if_backend_unavailable`, `_init_pg` in `C10dTorchCommsTestBase`, right?

File: test/distributed/test_c10d_torchcomms.py [hunk 3]
```diff
@@ -69,32 +72,32 @@ def _expected_reduce_result(self, op):
         raise ValueError(f"Unsupported op: {op}")
 
     @parametrize("op", REDUCE_OPS)
-    def test_allreduce(self, op):
+    def test_allreduce(self, device, op):
```
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py]: I perfer to use `device` parameter in here, and just like line 77, the tensor is created without specifing the `device` parameter, because the `torch.set_default_device(device)` is executed in `_init_pg` before creating the tensor, it would be better to remove those code and just create tensor with explicit `device` parameter.

File: test/distributed/test_c10d_torchcomms.py [hunk 4]
```diff
@@ -69,32 +72,32 @@ def _expected_reduce_result(self, op):
...
         self._skip_if_product_overflows(op)
         tensor = torch.tensor([self._rank_value], dtype=torch.float32)
         dist.all_reduce(tensor, op=op, group=self.pg)
         self.assertEqual(tensor.item(), self._expected_reduce_result(op))
 
-    def test_all_gather(self):
+    def test_all_gather(self, device):
```
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py]: ditto.

File: test/distributed/test_c10d_torchcomms.py
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py]: We should expand this one to support other device, right? And the device parameter can be leveraged to do that here.

Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:293]: This is cuda-specific, right?

Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:303]: cuda-specific, right?

File: test/distributed/test_c10d_torchcomms.py [hunk 5]
```diff
@@ -345,34 +347,34 @@ def _init_pg(cls, rank, world_size, rdvz_file):
...
+    def test_default_pg_has_cpu_backend(self, device):
         default_pg = dist.distributed_c10d._get_default_group()
         cpu_be = default_pg._get_backend(torch.device("cpu"))
         self.assertIsNotNone(cpu_be)
 
-    def test_default_pg_has_cuda_backend(self):
+    def test_default_pg_has_cuda_backend(self, device):
         default_pg = dist.distributed_c10d._get_default_group()
         cuda_be = default_pg._get_backend(torch.device("cuda"))
         self.assertIsNotNone(cuda_be)
```
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:371]: This one can be refactored to support other accelerator.

File: test/distributed/test_c10d_torchcomms.py [hunk 6]
```diff
@@ -345,34 +347,34 @@ def _init_pg(cls, rank, world_size, rdvz_file):
     def _rank_value(self):
         return self.rank + 1
 
-    def test_default_pg_has_cpu_backend(self):
+    def test_default_pg_has_cpu_backend(self, device):
         default_pg = dist.distributed_c10d._get_default_group()
```
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:364]: This one can be refactored to support other accelerator.

File: test/distributed/test_c10d_torchcomms.py
Comment (fffrog) [inline test/distributed/test_c10d_torchcomms.py:316]: The `AUtoQualify` can be refactor to support more accelerator.

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/distributed/test_c10d_torchcomms.py (hunk 2 above)
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:25]: Tracked in #194152 — a separate framework-level PR to `common_distributed.py`:
- `backend()` in both `DistributedTestBase` and `C10dTorchCommsTestBase` now calls `c10d.get_default_backend_for_device(device)` (with `"privateuse1"` alias resolution) instead of the hard-coded `if cuda/hpu/xpu ... else: gloo` ladder.
- `_skip_if_backend_unavailable` now has a `TORCHCOMM_HAS_HCCL` flag + `"hccl"` entry in `backend_flags`, so a device resolving to `hccl` skips cleanly when the backend is not built (instead of erroring with `No backend type associated with device type cpu`).
- `_init_pg` / `create_pg` device-setup guard generalized from `if "nccl" in backend or "xccl" in backend:` to `if backend != "gloo":` — gloo is the only CPU backend in `default_device_backend_map`; every other resolved backend is an accelerator comm backend requiring per-rank device setup.

File: test/distributed/test_c10d_torchcomms.py
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py]: `test_new_group_qualified_backend_passes_through` now uses `backend = self.backend(torch.device(device).type)` + `backend=f"{torch.device(device).type}:{backend}"`.

Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:303]: Confirmed NCCL-specific. Moved `test_new_group_with_pg_options` and `test_new_group_sequential_pg_options_produce_distinct_groups` to a new `TestC10dTorchCommsBasicCUDA` class (`only_for=["cuda"]`). Also generalized `test_new_group_bare_default_backend_is_auto_qualified` (bare `backend="nccl"` → `backend=self.backend(...)`) and removed the now-unused `_requires_cuda` helper.

File: test/distributed/test_c10d_torchcomms.py (hunk 6 above)
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:364]: `test_default_pg_has_cuda_backend` now uses `_get_backend(torch.device(torch.device(device).type))` — kept the original method name, generalized the implementation.

File: test/distributed/test_c10d_torchcomms.py
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:316]: `TestC10dTorchCommsInitAutoQualify` relabeled `CUDA` → `ACCELERATOR` with `allow_xpu=True`. `_init_pg` generalized: resolves `device_type` via `cls.device_type` (handling `callable` + `"privateuse1"` alias) → `cls.backend(device_type)`, uses `torch.get_device_module(device_type).set_device(rank)`. The NCCL-specific `test_new_group_nccl_lazy_builds_per_peer_group` (nccl-lazy is `devices=["cuda"]`) moved to a new `TestC10dTorchCommsInitAutoQualifyCUDA` class (`only_for=["cuda"]`).

File: test/distributed/test_c10d_torchcomms.py (hunk 2 above)
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:25]: Thanks for the detailed review!  see the replies below for each comment.

File: test/distributed/test_c10d_torchcomms.py
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:293]: see the reply below.

File: test/distributed/test_c10d_torchcomms.py (hunk 5 above)
Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py:371]: see the reply below.

File: test/distributed/test_c10d_torchcomms.py [hunk 7]
```diff
@@ -252,34 +249,55 @@ def test_new_group_delegates_to_split_group(self):
...
         ng = dist.new_group(ranks=ranks, backend=None)
         tensor = torch.tensor([self._rank_value], dtype=torch.float32)
         dist.all_reduce(tensor, group=ng)
         self.assertEqual(tensor.item(), sum(range(1, self.world_size + 1)))
 
-    def test_new_group_bare_default_backend_is_auto_qualified(self):
-        if self._requires_cuda():
+    def test_new_group_bare_default_backend_is_auto_qualified(self, device):
+        backend = self.backend(torch.device(device).type)
+        if backend == "gloo":
```
Comment (wjlFlyer) [inline test/distributed/test_c10d_torchcomms.py]: Torchcomms also support `gloo` backend.

Comment (twthighd) [reply test/distributed/test_c10d_torchcomms.py]: Good point — TorchComms does handle gloo as well. Removed the early-returns in both `test_new_group_bare_default_backend_is_auto_qualified` and `test_new_group_qualified_backend_passes_through`, so the cpu variant now exercises `new_group and the qualified `"cpu:gloo"` spelling, both going through TorchComms' `new_comm` path. Verified both pass on the cpu variant.

**Review by wjlFlyer — APPROVED** [review]
